2.9BSD/usr/man/cat1/cc.1

CC(1)               UNIX Programmer's Manual                CC(1)

NAME
     cc, pcc - C compiler

SYNOPSIS
     cc [ option ] ... file ...

     pcc [ option ] ... file ...

DESCRIPTION
     _C_c is the UNIX C compiler.  It accepts several types of
     arguments:

     Arguments whose names end with `.c' are taken to be C source
     programs; they are compiled, and each object program is left
     on the file whose name is that of the source with `.o' sub-
     stituted for `.c'.  The `.o' file is normally deleted, how-
     ever, if a single C program is compiled and loaded all at
     one go.

     In the same way, arguments whose names end with `.s' are
     taken to be assembly source programs and are assembled, pro-
     ducing a `.o' file.

     The following options are interpreted by _c_c.  See _l_d(1) for
     load-time options.

     -c      Suppress the loading phase of the compilation, and
             force an object file to be produced even if only one
             program is compiled.

     -p      Arrange for the compiler to produce code which
             counts the number of times each routine is called;
             also, if loading takes place, replace the standard
             startup routine by one which automatically calls
             _m_o_n_i_t_o_r(3) at the start and arranges to write out a
             _m_o_n._o_u_t file at normal termination of execution of
             the object program.  An execution profile can then
             be generated by use of _p_r_o_f(1).

     -f      In systems without hardware floating-point, use a
             version of the C compiler which handles floating-
             point constants and loads the object program with
             the floating-point interpreter.  Do not use if the
             hardware is present.

     -O      Invoke an object-code optimizer.

     -S      Compile the named C programs, and leave the
             assembler-language output on corresponding files
             suffixed `.s'.

     -P      Run only the macro preprocessor and place the result

Printed 7/22/83                                                 1

CC(1)               UNIX Programmer's Manual                CC(1)

             for each `.c' file in a corresponding `.i' file and
             has no `#' lines in it.

     -E      Run only the macro preprocessor and send the result
             to the standard output.  The output is intended for
             compiler debugging; it is unacceptable as input to
             _c_c.

     -V      Compile for use in overlaid programs.  This option
             is normally used with -c; the loading should be done
             directly with _l_d(1).  The special name _C__O_V_E_R_L_A_Y is
             automatically defined (see the -D option below).
             This is useful in header files that may be used both
             with and without overlays.

     -o _o_u_t_p_u_t
             Name the final output file _o_u_t_p_u_t.  If this option
             is used the file `a.out' will be left undisturbed.

     -D_n_a_m_e=_d_e_f
     -D_n_a_m_e  Define the _n_a_m_e to the preprocessor, as if by
             `#define'.  If no definition is given, the name is
             defined as 1.

     -U_n_a_m_e  Remove any initial definition of _n_a_m_e.

     -I_d_i_r   `#include' files whose names do not begin with `/'
             are always sought first in the directory of the _f_i_l_e
             argument, then in directories named in -I options,
             then in directories on a standard list.

     -B_s_t_r_i_n_g
             Find substitute compiler passes in the files named
             _s_t_r_i_n_g with the suffixes cpp, c0, c1 and c2.  If
             _s_t_r_i_n_g is empty, use a standard backup version.

     -t[p012]
             Find only the designated compiler passes in the
             files whose names are constructed by a -B option.
             In the absence of a -B option, the _s_t_r_i_n_g is taken
             to be `/usr/c/'.

     Other arguments are taken to be either loader option argu-
     ments, or C-compatible object programs, typically produced
     by an earlier _c_c run, or perhaps libraries of C-compatible
     routines.  These programs, together with the results of any
     compilations specified, are loaded (in the order given) to
     produce an executable program with name a.out.

     The major purpose of the `portable C compiler', _p_c_c, is to
     serve as a model on which to base other compilers.  _P_c_c does
     not support options -f, -V, -E, -B, and -t.  It provides, in

Printed 7/22/83                                                 2

CC(1)               UNIX Programmer's Manual                CC(1)

     addition to the language of _c_c, unsigned char type data and
     initialized bit fields.

FILES
     file.c              input file
     file.o              object file
     a.out               loaded output
     /tmp/ctm?           temporaries for _c_c
     /lib/cpp            preprocessor
     /lib/c[01]          compiler for _c_c
     /lib/fc[01]         floating-point compiler
     /lib/c2             optional optimizer
     /lib/crt0.o         runtime startoff
     /lib/mcrt0.o        startoff for profiling
     /lib/fcrt0.o        startoff for floating-point interpreta-
     tion
     /lib/libc.a         standard library, see _i_n_t_r_o(3)
     /usr/lib/libovc.a   standard library for use with overlays
     /usr/include        standard directory for `#include' files
     /tmp/pc*            temporaries for _p_c_c
     /usr/lib/ccom       compiler for _p_c_c

SEE ALSO
     B. W. Kernighan and D. M. Ritchie, _T_h_e _C _P_r_o_g_r_a_m_m_i_n_g
     _L_a_n_g_u_a_g_e, Prentice-Hall, 1978
     D. M. Ritchie, _C _R_e_f_e_r_e_n_c_e _M_a_n_u_a_l
     adb(1), ld(1), prof(1), monitor(3)

DIAGNOSTICS
     The diagnostics produced by C itself are intended to be
     self-explanatory.  Occasional messages may be produced by
     the assembler or loader.  Of these, the most mystifying are
     from the assembler, _a_s(1), in particular `m', which means a
     multiply-defined external symbol (function or data).

BUGS
     _P_c_c is little tried on the PDP11; specialized code generated
     for that machine has not been well shaken down.  The -O
     optimizer was designed to work with _c_c; its use with _p_c_c is
     suspect.  The optimizer is also known to (rarely) produce
     incorrect code with _c_c as well.

Printed 7/22/83                                                 3



2.9BSD/usr/src/cmd/cc.c
static	char	sccsid[] = "@(#)cc.c	2.6";	/*	SCCS id keyword	*/
#
# include <stdio.h>
# include <ctype.h>
# include <signal.h>
# include <whoami.h>

/* cc command */

# define MAXINC 10
# define MAXFIL 100
# define MAXLIB 100
# define MAXOPT 100
char	*tmp0;
char	*tmp1;
char	*tmp2;
char	*tmp3;
char	*tmp4;
char	*tmp5;
char	*outfile;
# define CHSPACE 1000
char	ts[CHSPACE+50];
char	*tsa = ts;
char	*tsp = ts;
char	*av[50];
char	*clist[MAXFIL];
char	*llist[MAXLIB];
int	pflag;
int	sflag;
int	cflag;
int	eflag;
int	iflag;
int	vflag;
int	exflag;
int	oflag;
int	proflag;
#ifdef MENLO_OVLY
int	ovlyflag;
	/*
	 * Predefined constant always true for an overlay compilation.
	 * Useful for sorting through include files...
	 */
#define	OVL_DEF	"-DC_OVERLAY"
#define OVLIBC	"-lovc"
#endif MENLO_OVLY
#define LD	"/bin/ld"
#define AS	"/bin/as"
#define LIBC	"-lc"
int	noflflag;
char	*chpass ;
char	*npassname ;
char	pass0[20] = "/lib/c0";
char	pass1[20] = "/lib/c1";
char	pass2[20] = "/lib/c2";
char	passp[20] = "/lib/cpp";
char	*pref = "/lib/crt0.o";
char	*copy();
char	*setsuf();
char	*strcat();
char	*strcpy();

main(argc, argv)
char *argv[]; 
{
	char *t;
	char *savetsp;
	char *assource;
	char **pv, *ptemp[MAXOPT], **pvt;
	int nc, nl, i, j, c, f20, nxo, na;
	int idexit();

	i = nc = nl = f20 = nxo = 0;
	setbuf(stdout, (char *)NULL);
	pv = ptemp;
	while(++i < argc) {
		if(*argv[i] == '-') switch (argv[i][1]) {
		default:
			goto passa;
		case 'S':
			sflag++;
			cflag++;
			break;
		case 'o':
			if (++i < argc) {
				outfile = argv[i];
				if ((c=getsuf(outfile))=='c'||c=='o') {
					error("Would overwrite %s", outfile);
					exit(8);
				}
			}
			break;
		case 'O':
			oflag++;
			break;
		case 'p':
			proflag++;
			break;
#ifdef MENLO_OVLY
		case 'V':
			ovlyflag++;
			*pv++ = OVL_DEF;
			if (pv >= ptemp+MAXOPT) {
				error("Too many DIUC options", (char *)NULL);
				--pv;
			}
			break;
#endif MENLO_OVLY
		case 'E':
			exflag++;
		case 'P':
			pflag++;
			*pv++ = argv[i];
		case 'c':
			cflag++;
			break;

		case 'f':
			noflflag++;
			if (npassname || chpass)
				error("-f overwrites earlier option", (char *)NULL);
			npassname = "/lib/f";
			chpass = "1";
			break;

		case '2':
			if(argv[i][2] == '\0')
				pref = "/lib/crt2.o";
			else {
				pref = "/lib/crt20.o";
				f20 = 1;
			}
			break;
		case 'D':
		case 'I':
		case 'U':
		case 'C':
			*pv++ = argv[i];
			if (pv >= ptemp+MAXOPT) {
				error("Too many DIUC options", (char *)NULL);
				--pv;
			}
			break;
		case 't':
			if (chpass)
				error("-t overwrites earlier option", (char *)NULL);
			chpass = argv[i]+2;
			if (chpass[0]==0)
				chpass = "012p";
			break;

		case 'B':
			if (npassname)
				error("-B overwrites earlier option", (char *)NULL);
			npassname = argv[i]+2;
			if (npassname[0]==0)
				npassname = "/usr/src/cmd/c/o";
			break;

		case 'v':
			vflag++;
			break;

		case 'i':
			iflag++;
			break;
		} 
		else {
passa:
			t = argv[i];
			if((c=getsuf(t))=='c' || c=='s'|| exflag) {
				clist[nc++] = t;
				if (nc>=MAXFIL) {
					error("Too many source files", (char *)NULL);
					exit(1);
				}
				t = setsuf(t, 'o');
			}
			if (nodup(llist, t)) {
				llist[nl++] = t;
				if (nl >= MAXLIB) {
					error("Too many object/library files", (char *)NULL);
					exit(1);
				}
				if (getsuf(t)=='o')
					nxo++;
			}
		}
	}
	if (npassname && chpass ==0)
		chpass = "012p";
	if (chpass && npassname==0)
		npassname = "/usr/src/cmd/c/";
	if (chpass)
		for (t=chpass; *t; t++) {
			switch (*t) {
			case '0':
				strcpy (pass0, npassname);
				strcat (pass0, "c0");
				continue;
			case '1':
				strcpy (pass1, npassname);
				strcat (pass1, "c1");
				continue;
			case '2':
				strcpy (pass2, npassname);
				strcat (pass2, "c2");
				continue;
			case 'p':
				strcpy (passp, npassname);
				strcat (passp, "cpp");
				continue;
			}
		}
	if (noflflag)
		pref = proflag ? "/lib/fmcrt0.o" : "/lib/fcrt0.o";
	else if (proflag)
		pref = "/lib/mcrt0.o";
	if(nc==0)
		goto nocom;
	if (pflag==0) {
		char FD;

		tmp0 = copy("/tmp/ctm0XXXXX");
		mktemp (tmp0);
		FD = creat (tmp0);
		if (FD < 0)
		{
			error("cc: cannot create temp", NULL);
			exit(1);
		}
		close (FD);
	}
	if (signal(SIGINT, SIG_IGN) != SIG_IGN)
		signal(SIGINT, idexit);
	if (signal(SIGTERM, SIG_IGN) != SIG_IGN)
		signal(SIGTERM, idexit);
	(tmp1 = copy(tmp0))[8] = '1';
	(tmp2 = copy(tmp0))[8] = '2';
	(tmp3 = copy(tmp0))[8] = '3';
	if (oflag)
		(tmp5 = copy(tmp0))[8] = '5';
	if (pflag==0)
		(tmp4 = copy(tmp0))[8] = '4';
	pvt = pv;
	for (i=0; i<nc; i++) {
		if (nc>1 || vflag)
			printf("%s:\n", clist[i]);
		if (getsuf(clist[i])=='s') {
			assource = clist[i];
			goto assemble;
		} 
		else
			assource = tmp3;
		if (pflag)
			tmp4 = setsuf(clist[i], 'i');
		savetsp = tsp;
		av[0] = "cpp";
		av[1] = clist[i];
		av[2] = exflag ? "-" : tmp4;
		na = 3;
		for(pv=ptemp; pv <pvt; pv++)
			av[na++] = *pv;
		av[na++]=0;
		if(vflag)
			doecho(passp, av);
		if (callsys(passp, av)) {
			cflag++;
			eflag++;
			continue;
		}
		av[1] = tmp4;
		tsp = savetsp;
		av[0]= "c0";
		if (pflag) {
			cflag++;
			continue;
		}
		na = 2;
		av[na++] = tmp1;
		av[na++] = tmp2;
		if (proflag)
			av[na++] = "-P";
#ifdef MENLO_OVLY
		if (ovlyflag)
			av[na++] = "-V";
#endif
		av[na++] = 0;
		if(vflag)
			doecho(pass0, av);
		if (callsys(pass0, av)) {
			cflag++;
			eflag++;
			continue;
		}
		av[0] = "c1";
		av[1] = tmp1;
		av[2] = tmp2;
		if (sflag)
			assource = tmp3 = setsuf(clist[i], 's');
		av[3] = tmp3;
		if (oflag)
			av[3] = tmp5;
		na = 4;
#ifdef MENLO_OVLY
		if (ovlyflag)
			av[na++] = "-V";
#endif
		av[na++] = 0;
		if(vflag)
			doecho(pass1, av);
		if(callsys(pass1, av)) {
			cflag++;
			eflag++;
			continue;
		}
		if (oflag) {
			av[0] = "c2";
			av[1] = tmp5;
			av[2] = tmp3;
			av[3] = 0;
			if(vflag)
				doecho(pass2, av);
			if (callsys(pass2, av)) {
				unlink(tmp3);
				tmp3 = assource = tmp5;
			} 
			else
				unlink(tmp5);
		}
		if (sflag)
			continue;
assemble:
		av[0] = "as";
		av[1] = "-u";
		na = 2;
#ifdef	MENLO_OVLY
		if (ovlyflag)
			av[na++] = "-V";
#endif	MENLO_OVLY
		av[na++] = "-o";
		av[na++] = setsuf(clist[i], 'o');
		av[na++] = assource;
		av[na++] = 0;
		cunlink(tmp1);
		cunlink(tmp2);
		cunlink(tmp4);
		if(vflag)
			doecho(AS, av);
		if (callsys(AS, av) > 1) {
			cflag++;
			eflag++;
			continue;
		}
	}
nocom:
	if (cflag==0 && nl!=0) {
		i = 0;
		av[0] = "ld";
		av[1] = "-X";
		av[2] = pref;
		j = 3;
		if (noflflag) {
			j = 4;
			if (iflag)
				av[3] = "-lfpsim_sep";
			else
				av[3] = "-lfpsim";
		}
		if (outfile) {
			av[j++] = "-o";
			av[j++] = outfile;
		}
		while(i<nl)
			av[j++] = llist[i++];
		if(f20)
			av[j++] = "-l2";
		else {
#ifdef MENLO_OVLY
			av[j++] = ovlyflag ? OVLIBC : LIBC;
#else
			av[j++] = LIBC;
#endif MENLO_OVLY
		}
		av[j++] = 0;
		if(vflag)
			doecho(LD, av);
		eflag |= callsys(LD, av);
		if (nc==1 && nxo==1 && eflag==0)
			cunlink(setsuf(clist[0], 'o'));
	}
	dexit();
}

idexit()
{
	eflag = 100;
	dexit();
}

dexit()
{
	if (!pflag) {
		cunlink(tmp1);
		cunlink(tmp2);
		if (sflag==0)
			cunlink(tmp3);
		cunlink(tmp4);
		cunlink(tmp5);
		cunlink(tmp0);
	}
	exit(eflag);
}

error(s, x)
char *s, *x;
{
	fprintf(exflag?stderr:stdout, s, x);
	putc('\n', exflag? stderr : stdout);
	cflag++;
	eflag++;
}




getsuf(as)
char as[];
{
	register int c;
	register char *s;
	register int t;

	s = as;
	c = 0;
	while(t = *s++)
		if (t=='/')
			c = 0;
		else
			c++;
	s -= 3;
	if (c<=14 && c>2 && *s++=='.')
		return(*s);
	return(0);
}

char *
setsuf(as, ch)
char *as;
{
	register char *s, *s1;

	s = s1 = copy(as);
	while(*s)
		if (*s++ == '/')
			s1 = s;
	s[-1] = ch;
	return(s1);
}

callsys(f, v)
char f[], *v[]; 
{
	int t, status;

	if ((t=fork())==0) {
		execv(f, v);
		printf("Can't find %s\n", f);
		exit(100);
	} else
		if (t == -1) {
			printf("Try again\n");
			return(100);
		}
	while(t!=wait(&status))
		;
	if (t = status&0377) {
		if (t!=SIGINT) {
			printf("Fatal error in %s\n", f);
			eflag = 8;
		}
		dexit();
	}
	return((status>>8) & 0377);
}

char *
copy(as)
char *as;
{
	char *malloc();
	register char *otsp, *s;

	otsp = tsp;
	s = as;
	while (*tsp++ = *s++)
		;
	if (tsp > tsa+CHSPACE) {
		tsp = tsa = malloc(CHSPACE+50);
		if (tsp==NULL) {
			error("no space for file names", (char *)NULL);
			dexit();
		}
	}
	return(otsp);
}

nodup(l, os)
char **l, *os;
{
	register char *t, *s;
	register int c;

	s = os;
	if (getsuf(s) != 'o')
		return(1);
	while(t = *l++) {
		while(c = *s++)
			if (c != *t++)
				break;
		if (*t=='\0' && c=='\0')
			return(0);
		s = os;
	}
	return(1);
}

cunlink(f)
char *f;
{
	if (f==NULL)
		return;
	unlink(f);
}

doecho(s, v)
char *s;
register char **v;
{
	printf ("%s", s);
	while (*v)
		printf (" %s", *v++);
	putchar ('\n');
	fflush (stdout);
}
